contracts/templates/ERC20Template3.sol

```diff
@@ -0,0 +1,1184 @@
+        require(is_valid_subscription(msg.sender), "No subscription");
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        return (agg_predvals_numer[slot], agg_predvals_denom[slot]);
+    }
+
+    function get_subscription_revenue_at_block(
+        uint256 blocknum
+    ) public view returns (uint256) {
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        return (subscription_revenue_at_block[slot]);
+    }
+
+    function get_prediction(
+        uint256 blocknum,
+        address predictoor
+    )
+        public
+        view
+        returns (Prediction memory prediction)
+    {
+        //allow predictoors to see their own submissions
+        require(msg.sender == predictoor || blocknum < block.number, "you shall not pass");
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        prediction = predobjs[slot][predictoor];
+    }
+
+    // ----------------------- MUTATING FUNCTIONS -----------------------
+
+    function submit_predval(
+        bool predval,
+        uint256 stake,
+        uint256 blocknum
+    ) external {
+        require(paused == false, "paused");
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        require(slot >= soonest_block_to_predict(block.number), "too late to submit");
+        require(!submitted_predval(slot, msg.sender), "already submitted");
+        
+        predobjs[slot][msg.sender] = Prediction(
+            predval,
+            stake,
+            msg.sender,
+            false
+        );
+
+        // update agg_predvals
+        agg_predvals_numer[slot] += stake * (predval ? 1 : 0);
+        agg_predvals_denom[slot] += stake;
+
+        emit PredictionSubmitted(msg.sender, slot, stake);
+        // safe transfer stake
+        IERC20(stake_token).safeTransferFrom(msg.sender, address(this), stake);
+    }
+
+    function payout_mul(
+        uint256[] calldata blocknums,
+        address predictoor_addr
+    ) external {
+      for (uint i = 0; i < blocknums.length; i++) {
+        payout(blocknums[i], predictoor_addr);
+      }
+    }
+
+    function payout(
+        uint256 blocknum,
+        address predictoor_addr
+    ) public nonReentrant {
+        require(submitted_predval(blocknum, predictoor_addr), "not submitted");
+        uint256 slot = rail_blocknum_to_slot(blocknum);
+        Prediction memory predobj = predobjs[slot][predictoor_addr];
+        if(predobj.paid) return; // just return if already paid, in order not to break payout_mul
+        
+        // if OPF hasn't submitted trueval in truval_submit_timeout blocks then cancel round
+        if (block.number > slot + truval_submit_timeout_block && truval_submitted[slot]==Status.Pending){
+            truval_submitted[slot]=Status.Canceled;
+        }
+
+        if(truval_submitted[slot]==Status.Pending){
+            // if Status is Pending, do nothing, just return
+            return; 
+        }
+        if(truval_submitted[slot]==Status.Canceled){
```

Comment:
There are three separate constructions of the event `PredictionPayout`. It makes this routine hard to follow.

Could we make this cleaner please?
- Idea 1: just create one event object at the top, with initial params. Then change its params a bit depending on each of the three scenarios. Then emit.
- Idea 2: other?

Reply:
unfortunately, contract size is 23.971 Kb.
Adding a couple of variables to store event params will make it > 24K, making it undeployable :(

Reply:
solved